File: tools/invariant-substrates/tally.py
#!/usr/bin/env python3
"""Aggregate invariant-substrate burn-down counts across skill.yaml files.

Reads every `.claude/skills/*/skill.yaml` and emits a markdown table of
per-skill counts plus a total row. The counts come from each file's
`counts:` block:

    counts:
      guess: N
      observed: N
      verified: N
      total: N

Purpose: make the "burn-down count is the honest backlog" promise in
`docs/INVARIANT-SUBSTRATES.md` machine-addressable. Every round, this
script prints a single table showing where the substrate portfolio
stands. A round where the `guess` total does not drop (and no new
substrates landed) is a calibration signal.

Deliberately stdlib-only — no PyYAML. The schema is simple enough
that line-grep is the right dependency level for v0.1. If the schema
grows (nested structures, arrays of counts, etc.) this script graduates
to PyYAML and the dependency gets declared.

Usage:
    python3 tools/invariant-substrates/tally.py
    python3 tools/invariant-substrates/tally.py --repo-root /path/to/repo
    python3 tools/invariant-substrates/tally.py --fail-on-no-progress  # CI mode
"""
from __future__ import annotations

import argparse
import glob
import os
import re
import sys
from dataclasses import dataclass
from typing import Optional


SKILL_RE = re.compile(r"^skill:\s*(\S+)\s*$")
SPEC_VERSION_RE = re.compile(r"^spec-version:\s*(\S+)\s*$")
SPEC_ROUND_RE = re.compile(r"^spec-filed-round:\s*(\S+)\s*$")
PORTABLE_RE = re.compile(r"^portable:\s*(true|false)\b")
COUNTS_START_RE = re.compile(r"^counts:\s*$")
# Count lines: two-space indent under `counts:` — the only place these
# keys appear at that indent in the current schema.
COUNT_LINE_RE = re.compile(r"^  (guess|observed|verified|total):\s*(\d+)\b")


@dataclass
class Substrate:
    path: str
    skill: Optional[str] = None
    spec_version: Optional[str] = None
    spec_round: Optional[str] = None
    portable: Optional[bool] = None
    guess: int = 0
    observed: int = 0
    verified: int = 0
    total: int = 0

    @property
    def declared_total_matches_sum(self) -> bool:
        return self.total == self.guess + self.observed + self.verified


def parse_substrate(path: str) -> Substrate:
    """Extract fields from a skill.yaml without a YAML parser.

    Works for the v0.1 schema. If the schema grows nested count blocks
    or multi-document YAML, this function graduates to PyYAML.
    """
    s = Substrate(path=path)
    in_counts = False
    with open(path, "r", encoding="utf-8") as fh:
        for raw in fh:
            line = raw.rstrip("\n")

            # Top-level scalar keys (column 0)
            if m := SKILL_RE.match(line):
                s.skill = m.group(1)
                in_counts = False
                continue
            if m := SPEC_VERSION_RE.match(line):
                s.spec_version = m.group(1)
                in_counts = False
                continue
            if m := SPEC_ROUND_RE.match(line):
                s.spec_round = m.group(1)
                in_counts = False
                continue
            if m := PORTABLE_RE.match(line):
                s.portable = (m.group(1) == "true")
                in_counts = False
                continue

            # `counts:` block
            if COUNTS_START_RE.match(line):
                in_counts = True
                continue

            # Any non-indented, non-empty, non-comment line ends the
            # counts block. Comments and blank lines inside are fine.
            if in_counts:
                if line and not line.startswith(" ") and not line.startswith("#"):
                    in_counts = False
                elif m := COUNT_LINE_RE.match(line):
                    value = int(m.group(2))
                    key = m.group(1)
                    setattr(s, key, value)

    return s


def find_substrates(repo_root: str) -> list[Substrate]:
    pattern = os.path.join(repo_root, ".claude", "skills", "*", "skill.yaml")
    paths = sorted(glob.glob(pattern))
    return [parse_substrate(p) for p in paths]


def format_report(subs: list[Substrate], repo_root: str) -> str:
    if not subs:
        return (
            "# Invariant-substrate tally\n\n"
            "No `skill.yaml` files found under `.claude/skills/`.\n"
        )

    lines: list[str] = []
    lines.append("# Invariant-substrate tally")
    lines.append("")
    lines.append(
        "Aggregated from every `.claude/skills/<name>/skill.yaml` under "
        f"`{os.path.basename(repo_root.rstrip('/')) or repo_root}`. "
        "Per-substrate tier counts plus portfolio totals. The `guess` column "
        "is the honest backlog — every entry there is a candidate for "
        "promotion to `observed` (at least one data point) or `verified` "
        "(mechanical check)."
    )
    lines.append("")
    lines.append(
        "| Skill | Spec ver | Round | Portable | Guess | Observed | Verified | Total | Sum-check |"
    )
    lines.append(
        "|---|---|---|---|---:|---:|---:|---:|---|"
    )

    tot_g = tot_o = tot_v = tot_t = 0
    portable_count = 0
    zeta_count = 0
    mismatch_count = 0
    for s in subs:
        portable_cell = (
            "yes" if s.portable is True
            else "no"  if s.portable is False
            else "-"
        )
        if s.portable is True:
            portable_count += 1
        elif s.portable is False:
            zeta_count += 1

        sum_ok = "ok" if s.declared_total_matches_sum else "MISMATCH"
        if not s.declared_total_matches_sum:
            mismatch_count += 1

        lines.append(
            f"| `{s.skill or '?'}` "
            f"| {s.spec_version or '-'} "
            f"| {s.spec_round or '-'} "
            f"| {portable_cell} "
            f"| {s.guess} "
            f"| {s.observed} "
            f"| {s.verified} "
            f"| {s.total} "
            f"| {sum_ok} |"
        )
        tot_g += s.guess
        tot_o += s.observed
        tot_v += s.verified
        tot_t += s.total

    lines.append(
        f"| **total ({len(subs)} substrates)** "
        f"| - | - | - "
        f"| **{tot_g}** | **{tot_o}** | **{tot_v}** | **{tot_t}** "
        f"| {'ok' if tot_t == tot_g + tot_o + tot_v else 'MISMATCH'} |"
    )
    lines.append("")
    lines.append(
        f"Portable / Zeta-specific split: **{portable_count} portable**, "
        f"**{zeta_count} zeta-specific**, {len(subs) - portable_count - zeta_count} unspecified."
    )
    if mismatch_count:
        lines.append("")
        lines.append(
            f"**{mismatch_count} substrate(s) have a declared `total` that does not match "
            "the sum of tiers.** Fix before relying on this report."
        )
    return "\n".join(lines) + "\n"


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--repo-root",
        default=os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..")),
        help="Path to the repo root (default: inferred from this script's location).",
    )
    parser.add_argument(
        "--fail-on-no-progress",
        action="store_true",
        help=(
            "Exit non-zero if the portfolio has any `guess` entries. "
            "Useful in CI after the posture matures; today the guess "
            "column is expected to be non-zero."
        ),
    )
    parser.add_argument(
        "--fail-on-mismatch",
        action="store_true",
        help="Exit non-zero if any substrate's declared total does not match the sum of tiers.",
    )
    args = parser.parse_args(argv)

    repo_root = os.path.abspath(args.repo_root)
    subs = find_substrates(repo_root)
    report = format_report(subs, repo_root)
    sys.stdout.write(report)

    total_guess = sum(s.guess for s in subs)
    mismatches = sum(1 for s in subs if not s.declared_total_matches_sum)

    if args.fail_on_mismatch and mismatches:
        return 2
    if args.fail_on_no_progress and total_guess > 0:
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: tools/invariant-substrates/tally.ts
#!/usr/bin/env bun
// tally.ts
//
// Aggregate invariant-substrate burn-down counts across every
// `.claude/skills/*/skill.yaml`. Emits a markdown table of
// per-skill counts plus a portfolio total row. Backs the
// "burn-down count is the honest backlog" promise in
// `docs/INVARIANT-SUBSTRATES.md`.
//
// Language choice: bun + TypeScript. Rationale + alternatives
// considered are recorded in
// `docs/DECISIONS/2026-04-20-tools-scripting-language.md`.
// Summary: bun+TS is the post-setup scripting default because
// (1) TypeScript is coming for UI anyway — runtime amortizes;
// (2) native Windows binary (no Git Bash slowness);
// (3) faction-hate sidestep (some contributors hate bash,
//     some hate PowerShell; this goes outside both);
// (4) static types on automation code;
// (5) cross-project consistency with SQLSharp.
// Decision held at medium confidence; watchlist in the ADR.
//
// Reads each `skill.yaml`'s top-level scalars and its
// `counts:` block:
//
//   skill: <name>
//   spec-version: <version>
//   spec-filed-round: <round>
//   portable: true|false
//
//   counts:
//     hypothesis: N      # formerly `guess`; both keys accepted
//     observed:   N      # during the round-43 rename
//     verified:   N
//     total:      N
//
// Both `hypothesis:` and `guess:` are accepted to keep the
// transition mechanical. A `skill.yaml` still using `guess:`
// is parsed but flagged for rename.
//
// Usage:
//   bun run tally:substrates                                 # markdown report
//   bun ./tools/invariant-substrates/tally.ts --fail-on-mismatch     # non-zero if declared total != sum of tiers
//   bun ./tools/invariant-substrates/tally.ts --fail-on-no-progress  # non-zero if any hypothesis entries remain

import { Glob } from "bun";
import { readFile } from "node:fs/promises";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const here = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(here, "..", "..");

type Flags = {
  failOnMismatch: boolean;
  failOnNoProgress: boolean;
  help: boolean;
};

function parseArgs(argv: readonly string[]): Flags {
  const flags: Flags = {
    failOnMismatch: false,
    failOnNoProgress: false,
    help: false,
  };
  for (const arg of argv) {
    switch (arg) {
      case "--fail-on-mismatch":
        flags.failOnMismatch = true;
        break;
      case "--fail-on-no-progress":
        flags.failOnNoProgress = true;
        break;
      case "-h":
      case "--help":
        flags.help = true;
        break;
      default:
        // unknown flag ignored, matching tally.sh behaviour
        break;
    }
  }
  return flags;
}

function printHelp(): void {
  const lines = [
    "tally.ts — aggregate invariant-substrate burn-down counts.",
    "",
    "Usage:",
    "  bun run tally:substrates                        # markdown report",
    "  bun ./tools/invariant-substrates/tally.ts --fail-on-mismatch",
    "  bun ./tools/invariant-substrates/tally.ts --fail-on-no-progress",
    "",
    "Reads every `.claude/skills/*/skill.yaml` and aggregates its",
    "top-level scalars and `counts:` block (hypothesis / observed /",
    "verified / total). `guess:` is accepted as a legacy alias for",
    "`hypothesis:` during the round-43 rename transition.",
  ];
  console.log(lines.join("\n"));
}

function stripComment(value: string): string {
  const hashAt = value.indexOf("#");
  return (hashAt === -1 ? value : value.slice(0, hashAt)).trim();
}

// Extract a top-level scalar. Matches at column 0 only, preserving
// the same discipline as the awk version so `counts.hypothesis` is
// never confused with a top-level `hypothesis` key.
function extractTopScalar(source: string, key: string): string | undefined {
  const pattern = new RegExp(`^${key}:\\s*(.*)$`, "m");
  const match = source.match(pattern);
  if (!match || match[1] === undefined) return undefined;
  const value = stripComment(match[1]);
  return value.length > 0 ? value : undefined;
}

// Extract a nested count under the `counts:` block. First match wins.
// The two-space indent is the only form this aggregator recognizes;
// it matches the shape the pilot skill.yaml files actually use.
function extractCount(source: string, key: string): number | undefined {
  const lines = source.split("\n");
  let inCounts = false;
  const nestedPattern = new RegExp(`^  ${key}:\\s*(.*)$`);
  for (const rawLine of lines) {
    if (/^counts:\s*$/.test(rawLine)) {
      inCounts = true;
      continue;
    }
    if (inCounts && /^[^\s#]/.test(rawLine)) {
      inCounts = false;
    }
    if (inCounts) {
      const match = rawLine.match(nestedPattern);
      if (match && match[1] !== undefined) {
        const value = stripComment(match[1]);
        const n = Number.parseInt(value, 10);
        return Number.isFinite(n) ? n : undefined;
      }
    }
  }
  return undefined;
}

type Row = {
  skill: string;
  specVer: string;
  specRound: string;
  portableCell: string;
  portableState: "yes" | "no" | "unspecified";
  hypothesis: number;
  observed: number;
  verified: number;
  total: number;
  sumCheck: "ok" | "MISMATCH";
  legacyKey: boolean;
};

async function collectRow(path: string): Promise<Row> {
  const source = await readFile(path, "utf8");

  const skill = extractTopScalar(source, "skill") ?? "?";
  const specVer = extractTopScalar(source, "spec-version") ?? "-";
  const specRound = extractTopScalar(source, "spec-filed-round") ?? "-";
  const portableRaw = extractTopScalar(source, "portable");

  let legacyKey = false;
  let hypothesis = extractCount(source, "hypothesis");
  if (hypothesis === undefined) {
    const guessFallback = extractCount(source, "guess");
    if (guessFallback !== undefined) {
      hypothesis = guessFallback;
      legacyKey = true;
    }
  }

  const h = hypothesis ?? 0;
  const o = extractCount(source, "observed") ?? 0;
  const v = extractCount(source, "verified") ?? 0;
  const t = extractCount(source, "total") ?? 0;

  let portableCell = "-";
  let portableState: Row["portableState"] = "unspecified";
  if (portableRaw === "true") {
    portableCell = "yes";
    portableState = "yes";
  } else if (portableRaw === "false") {
    portableCell = "no";
    portableState = "no";
  }

  const sum = h + o + v;
  const sumCheck: Row["sumCheck"] = t === sum ? "ok" : "MISMATCH";

  return {
    skill,
    specVer,
    specRound,
    portableCell,
    portableState,
    hypothesis: h,
    observed: o,
    verified: v,
    total: t,
    sumCheck,
    legacyKey,
  };
}

async function main(): Promise<void> {
  const flags = parseArgs(process.argv.slice(2));
  if (flags.help) {
    printHelp();
    return;
  }

  const glob = new Glob(".claude/skills/*/skill.yaml");
  const paths: string[] = [];
  // `dot: true` is required — `.claude` is a dotfile directory and Bun's
  // Glob otherwise skips it, returning zero matches. Discovered
  // 2026-04-20 against round-43 tally.sh parity.
  for await (const rel of glob.scan({ cwd: repoRoot, dot: true })) {
    paths.push(join(repoRoot, rel));
  }
  paths.sort();

  const rows = await Promise.all(paths.map(collectRow));

  const out: string[] = [];
  out.push("# Invariant-substrate tally\n");
  out.push(
    "Aggregated from every `.claude/skills/<name>/skill.yaml`. " +
      "Per-substrate tier counts plus portfolio totals. The " +
      "`hypothesis` column is the honest backlog — every entry " +
      "there is a candidate for promotion to `observed` (at " +
      "least one data point) or `verified` (mechanical check). " +
      "The tier names follow the " +
      "`docs/INVARIANT-SUBSTRATES.md` posture.\n",
  );
  out.push(
    "| Skill | Spec ver | Round | Portable | Hypothesis | Observed | Verified | Total | Sum-check | Legacy key? |",
  );
  out.push("|---|---|---|---|---:|---:|---:|---:|---|---|");

  let totH = 0;
  let totO = 0;
  let totV = 0;
  let totT = 0;
  let portableYes = 0;
  let portableNo = 0;
  let mismatches = 0;
  let legacyKeyHits = 0;

  for (const row of rows) {
    out.push(
      `| \`${row.skill}\` | ${row.specVer} | ${row.specRound} | ${row.portableCell} | ` +
        `${row.hypothesis} | ${row.observed} | ${row.verified} | ${row.total} | ` +
        `${row.sumCheck} | ${row.legacyKey ? "yes" : "no"} |`,
    );
    totH += row.hypothesis;
    totO += row.observed;
    totV += row.verified;
    totT += row.total;
    if (row.portableState === "yes") portableYes += 1;
    else if (row.portableState === "no") portableNo += 1;
    if (row.sumCheck === "MISMATCH") mismatches += 1;
    if (row.legacyKey) legacyKeyHits += 1;
  }

  const totalSum = totH + totO + totV;
  const grandCheck = totT === totalSum ? "ok" : "MISMATCH";

  out.push(
    `| **total (${rows.length} substrates)** | - | - | - | **${totH}** | **${totO}** | **${totV}** | **${totT}** | ${grandCheck} | - |`,
  );
  out.push("");
  out.push(
    `Portable / Zeta-specific split: **${portableYes} portable**, **${portableNo} zeta-specific**, ` +
      `${rows.length - portableYes - portableNo} unspecified.`,
  );

  if (legacyKeyHits > 0) {
    out.push("");
    out.push(
      `**${legacyKeyHits} substrate(s) still use the legacy \`guess:\` key.** ` +
        "Rename to `hypothesis:` per " +
        "`docs/DECISIONS/2026-04-20-tools-scripting-language.md` §Terminology.",
    );
  }

  if (mismatches > 0) {
    out.push("");
    out.push(
      `**${mismatches} substrate(s) have a declared \`total\` that does not match the sum of tiers.** ` +
        "Fix before relying on this report.",
    );
  }

  console.log(out.join("\n"));

  if (flags.failOnMismatch && mismatches > 0) {
    process.exit(2);
  }
  if (flags.failOnNoProgress && totH > 0) {
    process.exit(1);
  }
}

await main();
